blaze-client/src/main/scala/org/http4s/blaze/client/BlazeClient.scala

```diff
@@ -63,88 +61,78 @@ object BlazeClient {
+    response <- responseResource
+  } yield response
 
-        def loop: F[Resource[F, Response[F]]] =
-          borrow.use { next =>
-            val res: F[Resource[F, Response[F]]] = next.connection
-              .runRequest(req)
-              .adaptError { case EOF =>
-                new SocketException(s"HTTP connection closed: ${key}")
-              }
-              .map { (response: Resource[F, Response[F]]) =>
-                response.flatMap(r =>
-                  Resource.make(F.pure(r))(_ => manager.release(next.connection)))
-              }
+  private def prepareConnection(key: RequestKey): Resource[F, (A, F[TimeoutException])] = for {
+    conn <- borrowConnection(key)
+    responseHeaderTimeoutF <- addResponseHeaderTimeout(conn)
+  } yield (conn, responseHeaderTimeoutF)
 
-            responseHeaderTimeout match {
-              case responseHeaderTimeout: FiniteDuration =>
-                Deferred[F, Unit].flatMap { gate =>
-                  val responseHeaderTimeoutF: F[TimeoutException] =
-                    F.delay {
-                      val stage =
-                        new ResponseHeaderTimeoutStage[ByteBuffer](
-                          responseHeaderTimeout,
-                          scheduler,
-                          ec)
-                      next.connection.spliceBefore(stage)
-                      stage
-                    }.bracket(stage =>
-                      F.asyncF[TimeoutException] { cb =>
-                        F.delay(stage.init(cb)) >> gate.complete(())
-                      })(stage => F.delay(stage.removeStage()))
+  private def borrowConnection(key: RequestKey): Resource[F, A] =
+    Resource.makeCase(manager.borrow(key).map(_.connection)) {
+      case (conn, ExitCase.Canceled) =>
+        // Currently we can't just release in case of cancelation, beause cancelation clears the Write state of Http1Connection, so it migth result in isRecycle=true even if there's a half written request.
+        manager.invalidate(conn)
+      case (conn, _) => manager.release(conn)
+    }
 
-                  F.racePair(gate.get *> res, responseHeaderTimeoutF)
-                    .flatMap[Resource[F, Response[F]]] {
-                      case Left((r, fiber)) => fiber.cancel.as(r)
-                      case Right((fiber, t)) => fiber.cancel >> F.raiseError(t)
-                    }
-                }
-              case _ => res
-            }
-          }
+  private def addResponseHeaderTimeout(conn: A): Resource[F, F[TimeoutException]] =
+    responseHeaderTimeout match {
+      case d: FiniteDuration =>
+        Resource.apply(
+          Deferred[F, Either[Throwable, TimeoutException]].flatMap(timeout =>
+            F.delay {
+              val stage = new ResponseHeaderTimeoutStage[ByteBuffer](d, scheduler, ec)
+              conn.spliceBefore(stage)
+              stage.init(e => timeout.complete(e).toIO.unsafeRunSync())
```

Comment:
I'm actually not sure which one is better. My intuition says `.toIO.unsafeRunSync()` is more efficient when `F = IO`. and `runAsync` is more efficient otherwise. I'll change it to runAsync.